File: py/25.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None
class Stack:

    def __init__(self):
        self.stack = []

    def push(self, val):
        self.stack.append(val)

    def pop(self):
        return self.stack.pop()

    def top(self):
        return self.stack[-1]

    def empty(self):
        if len(self.stack) == 0:
            return 1
        else:
            return 0

    def depth(self) -> int:
        return len(self.stack)


class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:

        if k == 1:
            return head

        p = ListNode()
        h = ListNode()
        stack = Stack()
        fh = False

        while head != None:
            for i in range(0, k):
                stack.push(head)
                head = head.next
            if stack.depth() < k:
                break
            top = stack.top()
            while not stack.empty():
                p.next = stack.pop()
                p = p.next
                if not fh:
                    h.next = p
                    fh = True
            p.next = top
            p = p.next

        return h.next

File: cpp/25.cpp
#include <iostream>
#include <vector>
using namespace std;

struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution
{
public:
    ListNode *reverseKGroup(ListNode *head, int k)
    {
        if (head == NULL)
            return NULL;
        vector<ListNode *> p;
        for (int i = 0; i < k; i++)
        {
            ListNode *q = NULL;
            p.push_back(q);
        }
        ListNode *pi = head;
        ListNode *pre = NULL;
        ListNode *h = NULL;
        while (pi != NULL)
        {
            for (int i = 0; i < k; i++)
            {
                p[i] = pi;
                if (pi != NULL)
                    pi = pi->next;
            }
            if (h == NULL && p[k - 1] != NULL)
                h = p[k - 1];
            if (p[k - 1] != NULL)
            {
                if (pre != NULL)
                    pre->next = p[k - 1];
                pre = p[0];
                p[0]->next = p[k - 1]->next;
                for (int i = k - 1; i > 0; i--)
                    p[i]->next = p[i - 1];
            }
        }
        if (h == NULL)
            return head;
        else
            return h;
    }
};
